source/dub/commandline.d

```diff
@@ -77,6 +78,314 @@ CommandGroup[] getCommands()
+}
+
+/// It returns the first argument when it does not start with `-`
+unittest {
+	assert(extractCommandNameArgument(["test"]).value == "test");
+}
+
+/// There is nothing to extract when the arguments only contain the `test` cmd
+unittest {
+	assert(extractCommandNameArgument(["test"]).remaining == []);
+}
+
+/// It extracts two arguments when they are not a command
+unittest {
+	assert(extractCommandNameArgument(["-a", "-b"]).remaining == ["-a", "-b"]);
+}
+
+/// It returns the an empty string when it starts with `-`
+unittest {
+	assert(extractCommandNameArgument(["-test"]).value == "");
+}
+
+/** Handles the Command Line options and commands.
+*/
+struct CommandLineHandler
+{
+	/// The list of commands that can be handled
+	CommandGroup[] commandGroups;
+
+	/// General options parser
+	CommonOptions options;
+
+	/** Create the list of all supported commands
+
+	Returns:
+		Returns the list of the supported command names
+	*/
+	string[] commandNames()
+	{
+		return commandGroups.map!(g => g.commands.map!(c => c.name).array).join;
+	}
+
+	/** Parses the general options and sets up the log level
+		and the root_path
+	*/
+	void prepareOptions(CommandArgs args) {
+		LogLevel loglevel = LogLevel.info;
+
+		options.prepare(args);
+
+		if (options.vverbose) loglevel = LogLevel.debug_;
+		else if (options.verbose) loglevel = LogLevel.diagnostic;
+		else if (options.vquiet) loglevel = LogLevel.none;
+		else if (options.quiet) loglevel = LogLevel.warn;
+		else if (options.verror) loglevel = LogLevel.error;
+		setLogLevel(loglevel);
+
+		if (options.root_path.empty)
+		{
+			options.root_path = getcwd();
+		}
+		else
+		{
+			options.root_path = options.root_path.absolutePath.buildNormalizedPath;
+		}
+	}
+
+	/** Get an instance of the requested command.
+
+	If there is no command in the argument list, the `run` command is returned
+	by default.
+
+	If the `--help` argument previously handled by `prepareOptions`,
+	`this.options.help` is already `true`, with this returning the requested
+	command. If no command was requested (just dub --help) this returns the
+	help command.
+
+	Params:
+		name = the command name
+
+	Returns:
+		Returns the command instance if it exists, null otherwise
+	*/
+	Command getCommand(string name) {
+		if (name == "help" || (name == "" && options.help))
+		{
+			return new HelpCommand();
+		}
+
+		if (name == "")
+		{
+			name = "run";
+		}
+
+		foreach (grp; commandGroups)
+			foreach (c; grp.commands)
+				if (c.name == name) {
+					return c;
+				}
+
+		return null;
+	}
+
+	/** Get an instance of the requested command after the args are sent.
+
+	It uses getCommand to get the command instance and then calls prepare.
+
+	Params:
+		name = the command name
+		args = the command arguments
+
+	Returns:
+		Returns the command instance if it exists, null otherwise
+	*/
+	Command prepareCommand(string name, CommandArgs args) {
+		auto cmd = getCommand(name);
+
+		if (cmd !is null || cast(HelpCommand)cmd !is null)
```

Comment:
wrong code? if cmd is null the second part will always be `null !is null` and otherwise the second part won't be run.

You probably wanted to use
```suggestion
		if (cmd !is null && !(cast(HelpCommand)cmd))
```